discovery_engine_core/core/src/stack/filters.rs

```diff
@@ -38,13 +38,26 @@ impl ArticleFilter for DuplicateFilter {
         stack: &[Document],
         mut articles: Vec<Article>,
     ) -> Result<Vec<Article>, GenericError> {
-        let urls = history
+        let mut urls = history
             .iter()
-            .map(|doc| doc.url.as_str())
-            .chain(stack.iter().map(|doc| doc.resource.url.as_str()))
+            .map(|doc| Cow::from(doc.url.as_str()))
+            .chain(stack.iter().map(|doc| Cow::from(doc.resource.url.as_str())))
             .collect::<HashSet<_>>();
 
-        articles.retain(|article| !urls.contains(&article.link.as_str()));
+        let mut titles = history
+            .iter()
+            .map(|doc| Cow::from(&doc.title))
+            .chain(stack.iter().map(|doc| Cow::from(&doc.resource.title)))
+            .collect::<HashSet<_>>();
+
+        articles.retain(|article| {
+            let do_retain = !(urls.contains(&Cow::from(article.link.as_str()))
+                || titles.contains(&Cow::from(&article.title)));
+            urls.insert(Cow::from(article.link.to_string()));
+            titles.insert(Cow::from(article.title.clone()));
```

Comment:
with the example you gave this clones 2*100 strings just to filter out 7 articles. i get that we can't use the reference as it's not known outside the closure, so maybe we should address the issue in a different way: we could derive `Eq + Hash` for `Article` just for the `link` & `title` fields and pass around `HashSet<Article>` instead of `Vec<Article>` wherever we use articles. then the articles would already be unique and we can still use `HashSet::retain()` to filter them regarding the history & stack as before. wdyt?

Reply:
Deriving will not work because we want `link` or `title`. Manually deriving a case like this seems dangerous since this a requirement of this filter and not a property of `Article`.

Reply:
fair point, makes sense to keep the logic in the filter. maybe another way to avoid the excessive cloning?

Reply:
I have a couple of idea, let me try.

Reply:
No failed. The only way is to trade lot's of small allocation with a couple of big one and use your idea on `Eq + Hash`

```rust
       struct UniqueArticle(Article);                                                                   
                                                                                                         
        impl Eq for UniqueArticle {}                                                                     
        impl PartialEq for UniqueArticle {                                                               
            fn eq(&self, other: &Self) -> bool {                                                         
                self.0.link == other.0.link || self.0.title == other.0.title                             
            }                                                                                            
        }                                                                                                
                                                                                                         
        impl std::hash::Hash for UniqueArticle {                                                         
            fn hash<H: std::hash::Hasher>(&self, state: &mut H) {                                        
                self.0.link.hash(state);                                                                 
                self.0.title.hash(state);                                                                
            }                                                                                            
        }                                                                                                
                                                                                                         
        let urls: HashSet<_> = history                                                                   
            .iter()                                                                                      
            .map(|doc| doc.url.as_str())                                                                 
            .chain(stack.iter().map(|doc| doc.resource.url.as_str()))                                    
            .collect();                                                                                  
                                                                                                         
        let titles: HashSet<_> = history                                                                 
            .iter()                                                                                      
            .map(|doc| &doc.title)                                                                       
            .chain(stack.iter().map(|doc| &doc.resource.title))                                          
            .collect();                                                                                  
                                                                                                         
        let mut articles: HashSet<_> = articles.into_iter().map(UniqueArticle).collect();                
                                                                                                         
        articles.retain(|article| {                                                                      
            !(urls.contains(&article.0.link.as_str())                                                    
                || titles.contains(&&article.0.title))                                                   
        });                                                                                              
                                                                                                         
        Ok(articles.into_iter().map(|u| u.0).collect())   
```

Reply:
Up to you guys decide. I'm fine with both.

Reply:
I also had some other approaches but your approach seems to be the most reasonable one.

Through the way `Hash` is implemented lead it different hashes for equal types and through this inconsistencies. I switched it to a `BTreeSet` to avoid it.